### PR TITLE
[WASI-NN] Add ggml backend for llama

### PR DESCRIPTION
**Pull request by dm4**
- Add the ggml backend to wasi-nn plugin.
- Write entire model binary into a temp file and send the filename to ggml currently.

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/ggml.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+Expect<ErrNo> initExecCtx(WasiNNEnvironment &Env, uint32_t GraphId,
+                          uint32_t &ContextId) noexcept {
+  Env.NNContext.emplace_back(GraphId, Env.NNGraph[GraphId]);
+
+  ContextId = Env.NNContext.size() - 1;
+  return ErrNo::Success;
+}
+
+Expect<ErrNo> setInput(WasiNNEnvironment &Env, uint32_t ContextId,
+                       __attribute__((unused)) uint32_t Index,
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Fixed

File: plugins/wasi_nn/ggml.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  tempFile.close();
+
+  // Initialize ggml model.
+  gpt_params params;
+  params.model = modelfilepath;
+  llama_backend_init(params.numa);
+  std::tie(GraphRef.LlamaModel, GraphRef.LlamaContext) =
+      llama_init_from_gpt_params(params);
+  if (GraphRef.LlamaModel == nullptr) {
+    spdlog::error("[WASI-NN] Error: unable to load model.");
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Updated

File: plugins/wasi_nn/ggml.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    spdlog::error("[WASI-NN] Invalid ggml model.");
+    return ErrNo::InvalidArgument;
+  }
+
+  // TODO: pass the model directly to ggml
+  // Write ggml model to file.
+  std::string modelfilepath("ggml-model.bin");
+  std::ofstream tempFile(modelfilepath);
+  if (!tempFile) {
+    spdlog::error("[WASI-NN] Failed to create the temporary file.");
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Updated

File: plugins/wasi_nn/ggml.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  std::istringstream BinRead(BinModel);
+
+  if (BinModel.substr(0, 4) != "tjgg") {
+    spdlog::error("[WASI-NN] Invalid ggml model.");
+    return ErrNo::InvalidArgument;
+  }
+
+  // TODO: pass the model directly to ggml
+  // Write ggml model to file.
+  std::string modelfilepath("ggml-model.bin");
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Fixed

File: plugins/wasi_nn/ggml.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+#include "ggml.h"
+#include "wasinnenv.h"
+
+#ifdef WASMEDGE_PLUGIN_WASI_NN_BACKEND_GGML
+#include "ggml/common.h"
+#include "ggml/llama.h"
+#endif
+
+namespace WasmEdge::Host::WASINN::Ggml {
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Fixed

File: plugins/wasi_nn/ggml.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+#include <llama.h>
+#endif
+
+namespace WasmEdge::Host::WASINN::GGML {
+#ifdef WASMEDGE_PLUGIN_WASI_NN_BACKEND_GGML
+Expect<ErrNo> load(WasiNNEnvironment &Env, Span<const Span<uint8_t>> Builders,
+                   Device Device, uint32_t &GraphId) noexcept {
+  // The graph builder length must be 1.
+  if (Builders.size() != 1) {
+    spdlog::error("[WASI-NN] Wrong GraphBuilder Length {:d}, expect 1",
```
Comment (hydai) [inline plugins/wasi_nn/ggml.cpp:18]: Use `string_literal` for all const strings. This should be `"..."sv`.

File: plugins/wasi_nn/ggml.cpp [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+Expect<ErrNo> compute(WasiNNEnvironment &Env, uint32_t ContextId) noexcept {
+  auto &CxtRef = Env.NNContext[ContextId].get<Context>();
+  auto &GraphRef = Env.NNGraph[CxtRef.GraphId].get<Graph>();
+  if (CxtRef.LlamaInputs.size() == 0) {
+    spdlog::error("[WASI-NN] Llama input is not set!");
+    return ErrNo::InvalidArgument;
+  }
+
+  // Output start from prompt.
+  for (auto id : CxtRef.LlamaInputs) {
```
Comment (hydai) [inline plugins/wasi_nn/ggml.cpp:118]: `Id`.

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/ggml.cpp (hunk 6 above)
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Updated

File: plugins/wasi_nn/ggml.cpp [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    CxtRef.LlamaOutputs += llama_token_to_str(GraphRef.LlamaContext, id);
+  }
+
+  // Main predict loop.
+  // TODO: recompute a compressed context based on previous tokens once the
+  // cache is full.
+  const int max_context_size = llama_n_ctx(GraphRef.LlamaContext);
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Updated

File: thirdparty/ggml/common.cpp [hunk 9]
```diff
@@ -0,0 +1,701 @@
+#include "common.h"
```
[the remark this answers is not in the document]
Comment (dm4) [reply thirdparty/ggml/common.cpp:1]: Update the license info to LICENSE.spdx and add `thirdparty/ggml/LICENSE`.

File: plugins/wasi_nn/ggml.cpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+Expect<ErrNo> setInput(WasiNNEnvironment &Env, uint32_t ContextId,
+                       [[maybe_unused]] uint32_t Index,
+                       const TensorData &Tensor) noexcept {
+  auto &CxtRef = Env.NNContext[ContextId].get<Context>();
+  auto &GraphRef = Env.NNGraph[CxtRef.GraphId].get<Graph>();
+  std::string prompt(reinterpret_cast<char *>(Tensor.Tensor.data()),
+                     Tensor.Tensor.size());
+  CxtRef.LlamaInputs = llama_tokenize(GraphRef.LlamaContext, prompt, true);
+  const uint32_t max_context_size = llama_n_ctx(GraphRef.LlamaContext);
+  const uint32_t max_tokens_list_size = max_context_size - 4;
```
[the remark this answers is not in the document]
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: I believe the `-4` value here is reserved for specific tokens such as `<s>` and `</s>` to indicate the beginning and end of a sentence. I referred to [here](https://github.com/ggerganov/llama.cpp/blob/master/examples/simple/simple.cpp#L55-L56) for the implementation.

**Comment by dm4** [comment]
Update:
- Add license.
- Use `string_literal`.
- Correct naming style.
- Build and release WASI-NN ggml backend plugin.

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 11]
```diff
@@ -107,7 +107,6 @@ WasiNNLoadByName::bodyImpl(const Runtime::CallingFrame &Frame, uint32_t NamePtr,
   }
 
   // Get the name of model
-  std::vector<Span<uint8_t>> Builders;
```
Comment (hydai) [inline plugins/wasi_nn/wasinnfunc.cpp:110]: Why do we remove this? Is it not used anymore?

Comment (dm4) [reply plugins/wasi_nn/wasinnfunc.cpp:110]: Yes, I just help @CaptainVincent to remove this unused Builders.

**Comment by dm4** [comment]
> Do we have any plan for testing the ggml?

Yes, I'll add tests at `test/plugins/wasi_nn/wasi_nn.cpp`.

**Review by hydai — APPROVED** [review]
LGTM

**Pull request merged**